rewrite the loadbalancing document (#6689)
Kong/docs.konghq.com: app/_src/gateway/how-kong-works/load-balancing.md

---

Kong provides multiple ways of load balancing requests to multiple backend
services: the default DNS-based method, and an advanced set of load-balancing
algorithms using the Upstream entity.

The DNS load balancer is enabled by default and is limited to round-robin
load-balancing. The `upstream` entity has health-check and circuit-breaker
functionalities, besides the more advanced algorithms like least-connections,
consistent-hashing, and lowest-latency.

> Please read up on the [DNS caveats](#dns-caveats) depending on your infrastructure .

Reviewer: ```suggestion
Refer to the [DNS caveats](#dns-caveats) depending on your infrastructure .
```

## DNS-based load balancing

Every Service that has been defined with a `host` containing a hostname
(instead of an IP address) will automatically use DNS-based load balancing
if the name resolves to multiple IP addresses.

The DNS record `ttl` setting (time to live) determines how often the information
is refreshed. When using a `ttl` of 0, every request will be resolved using its
own DNS query. Obviously this will have a performance penalty, but the latency of
updates/changes will be very low.

> The (weighted) round-robin algorithm used depends on the DNS record type of the

Reviewer: ```suggestion
The weighted round-robin algorithm used depends on the DNS record type of the
```

Author: replaced this one slightly different, please have a look.

> The (weighted) round-robin algorithm used depends on the DNS record type of the

Author: ```suggestion
The round-robin algorithm used, weighted or not, depends on the DNS record type of the
```

randomized. So when using records with a `ttl` of 0, the nameserver is
expected to randomize the record entries.

## Advanced load-balancing

Advanced load-balancing algorithms are available through the `upstream` entity.

> When using these load-balancers, the adding and removing of backend services will

Reviewer: ```suggestion
When using these load balancers, the adding and removing of backend services will
```

be handled by Kong, and no DNS updates will be necessary. Kong will act as the
service registry.

> Configuring the load-balancers is done through the `upstream` and `target`

Reviewer: ```suggestion
Configuring the load balancers is done through the `upstream` and `target`
```

Would it resolve to an SRV record, then also the `port` and `weight` fields
from the DNS record would be picked up, and would overrule the given port `123`
and `weight=100`.

> __NOTE__: similar to the DNS based load-balancing, only the highest priority

Reviewer: ```suggestion
{:.note}
> **Note**: similar to the DNS based load-balancing, only the highest priority
```

entries (the lowest values) in an SRV record will be used.

The balancer will honor the DNS record's `ttl` setting, upon expiry it queries the
nameserver and updates the balancer.

> __Exception__: When a DNS record has `ttl=0`, the hostname will be added

Reviewer: ```suggestion
{:.important}
> **Exception**: When a DNS record has `ttl=0`, the hostname will be added
```

as a single target, with the specified weight. Upon every proxied request
to this target it will query the nameserver again.

### Fighting load-balancers

As described in the [target](#target) paragraph, the targets can be specified as hostnames.
In orchestrated environments like k8s or docker-compose, the IP addresses and ports
are mostly ephemeral and SRV records must be used to find the appropriate backends and
to stay up to date.

On a DNS level many infrastructure tools can also provide load-balancing type features.
These are mostly service-discovery tools that will have their own health-checks and
will randomize DNS records, or only return a small subset of available peers.

The Kong load balancers and the DNS based tools often fight each other. The nameserver will
provide as little information as possible to force clients to follow its scheme, where

> Kong tries to get all backends to properly set up its load balancers (and health-checks).

Reviewer: ```suggestion
Kong tries to get all backends to properly set up its load balancers and health-checks.
```

> To make sure that Kong can properly do its job, check that;

Reviewer: ```suggestion
In your environment, ensure that:
```

- the nameserver sets the truncation flag on the responses when it cannot fit all

>   records in the UDP response (this will force Kong to retry using TCP)

Reviewer: ```suggestion
  records in the UDP response. This will force Kong to retry using TCP. 
```

> - allow TCP queries on the nameserver.

Reviewer: ```suggestion
- TCP queries are allowed on the nameserver.
```

## Balancing algorithms

> The load-balancers support the following load balancing algorithms:

Reviewer: ```suggestion
The load balancers support the following load-balancing algorithms:
```

* `round-robin`
* `consistent-hashing`
* `least-connections`
{% if_version gte:3.2.x %}
* `latency`
{% endif_version %}

These algorithms are only available when using the `upstream` entity, see
[Advanced load-balancing](#advanced-load-balancing).

> __NOTE__: for all these algorithms it is important to understand how the weights

Reviewer: ```suggestion
{:.note}
> **Note**: for all these algorithms it is important to understand how the weights
```

and ports of the individual backends are being set up. See the [Target](#target)
paragraph on how the actual weights and ports are being determined based on user
configuration as well DNS results.

### Round-Robin

> The Round Robin algorithm will be done in a weighted manner. It will be identical

Reviewer: ```suggestion
The round-robin algorithm will be done in a weighted manner. It will be identical
```

in results to the DNS based load-balancing, but due to it being an `upstream`
the additional features for health-checks and circuit-breakers will be available
in this case.

> #### Considerations
> 

Reviewer: ```suggestion
When choosing this algorithm, consider the following:
```

- good distribution of requests.
- fairly static, as only DNS updates or `target` updates can influence the
  distribution of traffic.

> - does not improve cache-hit ratio's.

Reviewer: ```suggestion
- does not improve cache-hit ratios.
```

### Consistent-Hashing

With the consistent-hashing algorithm a configurable client-input will be used to
calculate a hash-value. This hash-value will then be tied to a specific backend
server.

A common example would be to use the `consumer` as a hash-input. Since this ID is
the same for every request from that user, it will ensure that the same user will
consistently be dealt with by the same backend server. This will allow for cache
optimizations on the backend, since each of the servers only serves a fixed subset
of the users, and hence can improve its cache-hit-ratio for user related data.

This algorithm implements the [ketama principle](https://github.com/RJ/ketama) to

> maximize hashing stability (minimize consistency loss) upon changes to the list

Reviewer: ```suggestion
maximize hashing stability and minimize consistency loss upon changes to the list
```

  specified in the `hash_on_cookie_path` field. If the specified cookie is not
  present in the request, it will be set by the response. Hence, the `hash_fallback`
  setting is invalid if `cookie` is the primary hashing mechanism.
  The generated cookie will have a random UUID value. So the first assignment will
  be random, but then sticks because it is preserved in the cookie.

The consistent-hashing balancer is designed to work both with a single node as well
as in a cluster. When using the hash based algorithm it is important that all nodes
build the exact same balancer-layout to make sure they all work identical. To do

> this the balancer must be build in a deterministic way (see DNS considerations below).

Reviewer: ```suggestion
this the balancer must be built in a deterministic way. 

When choosing this algorithm, consider the following: 
```

> #### Considerations
> 

Reviewer: ```suggestion

```

> - improves backend cache-hit ratio's.

Reviewer: ```suggestion
- improves backend cache-hit ratios.
```

> - requires enough cardinality in the hash-inputs to distribute evenly. Eg. hashing on
>   a header that only has 2 possible values does not make sense.

Reviewer: ```suggestion
- requires enough cardinality in the hash-inputs to distribute evenly (for example, hashing on
  a header that only has 2 possible values does not make sense).
```

- the cookie based approach will work well for browser based requests, but less so
  for machine-2-machine clients which will often omit the cookie.
- avoid using hostnames in the balancer as the
  balancers might/will slowly diverge because the DNS ttl has only second precision
  and renewal is determined by when a name is actually requested. On top of this is
  the issue with some nameservers not returning all entries, which exacerbates
  this problem. So when using the hashing approach in a Kong cluster, preferably add
  `target` entities by their IP address. This problem can be mitigated by balancer
  rebuilds and higher ttl settings. 

### Least-Connections

This algorithm keeps track of the number of in-flight requests for each backend.
The weights are used to calculate "connection-capacity" of a backend. Requests are

> routed towards the backend with the highest spare capacity.
> 
> #### Considerations
> 

Reviewer: ```suggestion
routed towards the backend with the highest spare capacity.

When choosing this algorithm, consider the following:
```

- good distribution of traffic.
- does not improve cache-hit ratio's.
- more dynamic since slower backends will have more connections open, and hence
  new requests will be routed to other backends automatically.

{% if_version gte:3.2.x %}
### Latency

The `latency` algorithm is based on peak EWMA (exponentially weighted moving average),
which ensures that the balancer selects the backend by lowest latency
(`upstream_response_time`). The latency metric used is the full request cycle, from
TCP connect to body response time. Since it is a moving average, the metrics will
"decay" over time.

Weights will not be taken into account.

> #### Considerations
> 

Reviewer: ```suggestion
When choosing this algorithm, consider the following:
```

- good distribution of traffic provided there is enough base-load to keep the

>   metrics alive (since they are "decaying")

Reviewer: ```suggestion
  metrics alive, since they are "decaying".
```

- not suitable for long-lived connections like websockets or server-sent events (SSE)
- very dynamic since it will constantly optimize.

> - ideally this works best with low variance in latencies. This means mostly similar
>   shaped traffic and even workloads for the backends. For example usage
>   with a GraphQL backend serving small-fast queries as well big-slow ones will result
>   in high variance in the latency metrics, which will skew the metrics.

Reviewer: ```suggestion
- ideally, this works best with low variance in latencies. This means mostly similar
  shaped traffic and even workloads for the backends. For example, usage
  with a GraphQL backend serving small-fast queries as well big-slow ones will result
  in high variance in the latency metrics, which will skew the metrics.
```

> - properly set up the backend capacity and ensure proper network latency to prevent
>   resource starvation. For example; 2 servers, one a small capacity close by (low
>   network latency), the other high capacity far away (high latency). Most traffic
>   will be routed to the small one, until its latency starts going up. The latency
>   going up however means the small server is most likely suffering from resource
>   starvation. So in this case the algorithm will keep the small server in a constant
>   state of resource starvation, which is most likely not efficient.

Reviewer: ```suggestion
- properly set up the backend capacity and ensure proper network latency to prevent
  resource starvation. For example, use 2 servers: one a small capacity close by (low
  network latency), the other high capacity far away (high latency). Most traffic
  will be routed to the small one, until its latency starts going up. The latency
  going up however means the small server is most likely suffering from resource
  starvation. So, in this case, the algorithm will keep the small server in a constant
  state of resource starvation, which is most likely not efficient.
```
